pkg/bpf/bpffs.go

```diff
@@ -104,100 +123,226 @@ func OpenAfterMount(m *Map) error {
 	return nil
 }
 
+// removeMap removes a map from the BPF filesystem.
+func removeMap(name string) {
+	os.Remove(MapPath(name))
+}
+
+// RemoveAfterMount schedules a map to be deleted after the BPF filesystem has
+// been mounted. If the filesystem is already mounted, the map is deleted
+// immediately.
+func RemoveAfterMount(name string) {
+	mountMutex.Lock()
+	defer mountMutex.Unlock()
+
+	if mounted {
+		removeMap(name)
+		return
+	}
+
+	log.WithField(logfields.BPFMapName, name).Debug("bpffs is not mounted yet; adding to list of maps to remove once it is mounted")
+	delayedRemoves = append(delayedRemoves, name)
```

Comment:
I would be surprised if this case needs to be so complex. AFAIK it handles upgrades from versions of Cilium with an old map name, which can never be the case if BPFFS is not yet mounted. Did you observe an issue with the previous `os.Remove()` and ignore the error behaviour?

Reply:
Yes, I observed a serious problem with the previous `os.Remove()` usage. It simply **never worked** (at least at time when I was writing my patches), because it tried to remove the map file before BPFFS was mounted. You need to mount BPFFS before you want to remove an old map or any file from mapRoot.

>AFAIK it handles upgrades from versions of Cilium with an old map name, which can never be the case if BPFFS is not yet mounted.

It can be the case,  when the mountpoint is `/sys/fs/bpf` and we have one of those situations:
- BPFFS is mounted on host, so BPFFS tree lifecycle doesn't depend on container's lifecycle and has the same files between Cilium restarts
- Cilium is not containerized, runs as a systemd service on host, and BPFFS is specified in fstab

In both situations, `/sys/fs/bpf` content doesn't "refresh" between Cilium restarts. So, in those cases, if you upgrade Cilium and restart it, you still have the old files, so removing the old unused ones is a good approach.

Reply:
The previous usage of `os.Remove()` is here, in `init()` function:

https://github.com/cilium/cilium/blob/master/pkg/maps/tunnel/tunnel.go#L63

Reply:
You're right, I will dismiss that change.